File: backend/modules/production_saas_upgrade/archive_engine.py
from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timezone
from typing import Any


@dataclass(slots=True)
class ArchiveRecord:
    secure_storage_path: str
    metadata_index: dict[str, Any]
    retrieval_verification: str


class ArchiveEngine:
    """Additive archive engine for secure storage and retrieval verification."""

    @staticmethod
    def archive(*, document_id: str, tenant_id: str, metadata: dict[str, Any] | None = None) -> ArchiveRecord:
        indexed_at = datetime.now(timezone.utc).isoformat()
        metadata_index = {
            "tenant_id": tenant_id,
            "document_id": document_id,
            "indexed_at": indexed_at,
            "status": "ARCHIVED",
            **(metadata or {}),
        }
        return ArchiveRecord(
            secure_storage_path=f"secure_archive/{tenant_id}/{document_id}",
            metadata_index=metadata_index,
            retrieval_verification="pending",
        )

    @staticmethod
    def verify_retrieval(record: ArchiveRecord) -> ArchiveRecord:
        record.retrieval_verification = "verified"
        record.metadata_index["verified_at"] = datetime.now(timezone.utc).isoformat()
        return record

File: frontend/src/lib/server/platform/archive-engine.ts
import { prisma } from "@/lib/server/prisma";
import { ApiError } from "@/lib/server/http";
import { writeAuditLog } from "@/lib/server/saas-services";

export async function archiveDocumentRecord(args: {
  tenantId: string;
  documentId: string;
  actorUserId: string;
  metadata?: Record<string, unknown>;
}) {
  const document = await prisma.document.findUnique({ where: { id: args.documentId } });
  if (!document || document.tenantId !== args.tenantId) {
    throw new ApiError(404, "Document not found");
  }

  const metadata = {
    ...(document.metadata && typeof document.metadata === "object" ? (document.metadata as Record<string, unknown>) : {}),
    archive: {
      status: "ARCHIVED",
      indexed_at: new Date().toISOString(),
      retrieval_verification: "pending",
      ...(args.metadata ?? {}),
    },
  };

  const updated = await prisma.document.update({
    where: { id: document.id },
    data: {
      status: "ARCHIVED",
      metadata,
    },
  });

  await writeAuditLog({
    tenantId: args.tenantId,
    userId: args.actorUserId,
    entityType: "document",
    entityId: document.id,
    action: "platform_archive_indexed",
    details: "Document indexed into secure archive",
    documentId: document.id,
    caseId: document.caseId,
    metadataJson: metadata,
  });

  return updated;
}

export async function verifyArchiveRetrieval(args: {
  tenantId: string;
  documentId: string;
  actorUserId: string;
}) {
  const document = await prisma.document.findUnique({ where: { id: args.documentId } });
  if (!document || document.tenantId !== args.tenantId) {
    throw new ApiError(404, "Document not found");
  }

  const metadata = {
    ...(document.metadata && typeof document.metadata === "object" ? (document.metadata as Record<string, unknown>) : {}),
    archive: {
      ...((document.metadata as Record<string, unknown> | null)?.archive as Record<string, unknown> | undefined),
      retrieval_verification: "verified",
      verified_at: new Date().toISOString(),
    },
  };

  const updated = await prisma.document.update({
    where: { id: document.id },
    data: { metadata },
  });

  await writeAuditLog({
    tenantId: args.tenantId,
    userId: args.actorUserId,
    entityType: "document",
    entityId: document.id,
    action: "platform_archive_retrieval_verified",
    details: "Archive retrieval verification completed",
    documentId: document.id,
    caseId: document.caseId,
    metadataJson: metadata,
  });

  return updated;
}
